Mathlib/GroupTheory/Perm/Centralizer.lean

```diff
@@ -0,0 +1,642 @@
+  ext x
+  simp only [mul_apply, toCentralizer_apply]
+  by_cases hx : x ∈ (c : Perm α).support
+  · rw [(mem_cycleFactorsFinset_iff.mp c.prop).2 x hx]
+    have := newK_commute_zpow_apply a τ x 1
+    simp only [zpow_one] at this
+    rw [this, ← (mem_cycleFactorsFinset_iff.mp ((τ : Perm g.cycleFactorsFinset) c).prop).2]
+    rw [newK_apply_mem_support_cycle_iff]
+    exact hx
+  · rw [not_mem_support.mp hx, eq_comm, ← not_mem_support, newK_apply_mem_support_cycle_iff]
+    exact hx
+
+theorem toPermHom_apply_toCentralizer :
+    (toPermHom g) (toCentralizer a τ) = (τ : Perm g.cycleFactorsFinset) := by
+  apply ext
+  intro c
+  rw [OnCycleFactors.toPermHom_apply, toCentralizer_equivariant]
+
+end Basis
+
+namespace OnCycleFactors
+
+open Basis BigOperators Nat Equiv.Perm Equiv Subgroup
+
+theorem mem_range_toPermHom_iff {τ} : τ ∈ (toPermHom g).range ↔
+    ∀ c, (τ c : Perm α).support.card = (c : Perm α).support.card := by
+  constructor
+  · rintro ⟨k, rfl⟩ c
+    rw [coe_toPermHom, Equiv.Perm.support_conj]
+    apply Finset.card_map
+  · obtain ⟨a⟩ := Basis.nonempty g
+    exact fun hτ ↦ ⟨toCentralizer a ⟨τ, hτ⟩, toPermHom_apply_toCentralizer a ⟨τ, hτ⟩⟩
+
+/-- Unapplied variant of `Equiv.Perm.mem_range_toPermHom_iff` -/
+theorem mem_range_toPermHom_iff' {τ} : τ ∈ (toPermHom g).range ↔
+    (fun (c : g.cycleFactorsFinset) ↦ (c : Perm α).support.card) ∘ τ =
+      fun (c : g.cycleFactorsFinset) ↦ (c : Perm α).support.card := by
+  rw [mem_range_toPermHom_iff, funext_iff]
+  simp only [Finset.coe_sort_coe, Subtype.forall, Function.comp_apply]
+
+/-- Computes the range of `Equiv.Perm.toPermHom g` -/
+theorem range_toPermHom_eq_range_toPermHom' :
+    (toPermHom g).range = range_toPermHom' g := by
+  ext τ
+  rw [mem_range_toPermHom_iff, mem_range_toPermHom'_iff]
+
+theorem nat_card_range_toPermHom :
+    Nat.card (toPermHom g).range =
+      ∏ n in g.cycleType.toFinset, (g.cycleType.count n)! := by
+  classical
+  let sc (c : g.cycleFactorsFinset) : ℕ := (c : Perm α).support.card
+  suffices Fintype.card (toPermHom g).range =
+    Fintype.card { k : Perm g.cycleFactorsFinset | sc ∘ k = sc } by
+    simp only [Nat.card_eq_fintype_card, this, Set.coe_setOf,
+      DomMulAct.stabilizer_card', ← CycleType.count_def]
+    apply Finset.prod_congr _ (fun _ _ => rfl)
+    ext n
+    simp only [Finset.univ_eq_attach, Finset.mem_image, Finset.mem_attach,
+        sc, true_and, Subtype.exists, exists_prop, Multiset.mem_toFinset]
+    simp only [cycleType_def, Function.comp_apply, Multiset.mem_map, Finset.mem_val]
+  simp only [← SetLike.coe_sort_coe, Fintype.card_eq_nat_card]
+  congr
+  ext
+  rw [SetLike.mem_coe, mem_range_toPermHom_iff']
+  simp
+
+section Kernel
+/- Here, we describe the kernel of `g.OnCycleFactors.toPermHom` -/
+
+open BigOperators Nat OnCycleFactors Subgroup
+
+variable (g) in
+/-- The parametrization of the kernel of `toPermHom` -/
+def θHom : (Perm (Function.fixedPoints g)) ×
```

Comment:
This name does not seem very descriptive. But I don't think this is a private def, right?
Would something like `kerParam` sound good to you?

Reply:
This name and the previous one were given by lack of idea. To be honest, I was hoping that the issue would be raised during the review process.

Reply:
What do you think of the name `kerParam` that I suggested?